---
title: "R crash course / 05"
author: ""
date: "`r Sys.Date()`"
output:
  ioslides_presentation: 
    widescreen: true
    smaller: true
    css: files/style.css
    logo: files/logo_bihealth_en.png
---

<style>
.forceBreak { -webkit-column-break-after: always; break-after: column; }
</style>

```{r,echo=FALSE,warning=FALSE,message=FALSE}
## Set default options for the knitr RMD processing
knitr::opts_chunk$set(echo=TRUE,eval=FALSE,warning=FALSE,message=FALSE,fig.width=5,fig.height=5,cache=FALSE,autodep=TRUE, results="hide")
knitr::opts_knit$set(root.dir = "../")
```

## Recap for yesterday

## Aims for today

* Creating your first report -> bringing together everything you learned
* Introduction into statistics with R
* Visualizing data
* Your topics!

# Rmarkdown / Quarto

## Creating your first report - using Rmarkdown

 * Creating pretty tables -> Rmarkdown
    * [Video: introduction R markdown](https://youtu.be/vBeGakSENi8) (15 min)
    * [Video: basic document formatting in Rmarkdown documents](https://youtu.be/rfMTn4NBDkQ) (9 min)
    * [optional: [Video: including references in Rmarkdown documents](https://youtu.be/DUG1tEL5kXI) (7 min) ]
    * Nice tables with [kableExtra](https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html)

## Quarto vs Rmarkdown vs Markdown

 * Markdown is a lightweight markup language with plain text formatting
   syntax. This has nothing to do with R, it is about formatting only.
 * Rmarkdown is a variant of Markdown that allows to include R code
   that is executed and the results are included in the document.
 * Quarto is a new document format that is based on Rmarkdown (and
   backwards compatible) but has more features, looks nicer and is more
   flexible. Quarto works with other languages as well, not only R.

## The iris data set  {.columns-2 .smaller}

  * Measurements of flowers of three species of iris
  * 150 observations, 4 measurements each (sepal and petal width and length)
  * Early machine learning (linear discriminant analysis)

*Fisher RA. The use of multiple measurements in taxonomic problems. Annals
of Eugenics. 1936 Sep;7(2):179-88.*

<p class="forceBreak"></p>

![Iris versicolor](images/1280px-Iris_versicolor_3.jpg){ width=300px }

## Generating tables from data frames

Simple tables:

```{r eval=TRUE,results="markdown"}
data(iris) # not necessary

# generate table (idiom):
knitr::kable(iris[ 1:10, ])
```

## Generating tables from data frames

Nicely formatted tables:


```{r eval=TRUE,results="markdown"}

# generate table (idiom):
pander::pander(iris[ 1:5, ])
```

(this looks much better in a Word document!)

## Generating tables from data frames

Interactive tables:

```{r eval=TRUE,results="markdown"}
library(DT)

# generate table (idiom):
datatable(iris, options = list(pageLength = 5))
```

## Exercise 5.1

 * Create a new Rmarkdown document
 * Add a title, author, date
 * Add a table with the iris dataset
 * Add a plot with the iris dataset
 * Add a reference to the ggplot2 package
 * Knit the document to HTML

# Simple statistics with R

## Simple statistics (demo)

 * descriptive statistics
 * t.test
 * PCA

## Descriptive statistics

 * `summary` function(s)
 * `summary_colorDF` function
 * `mean`, `sd`, `var`, `median`, `quantile`, `range`, `min`, `max`

## Descriptive statistics

`tapply(x, groups, func)` applies a function `func` to some data `x`
grouped by `groups`.

```{r, eval=TRUE,results="markdown"}
library(tidyverse)
labwide <- read_csv("Datasets/labresults_wide_full.csv")[, -1 ]
mean(labwide$NEUT)
sd(labwide$NEUT)
```

## Descriptive statistics

`tapply(x, groups, func)` applies a function `func` to some data `x`
grouped by `groups`.


```{r, eval=TRUE,results="markdown"}
tapply(labwide$NEUT, labwide$ARM, mean)
tapply(labwide$NEUT, labwide$Timepoint, mean)
tapply(labwide$NEUT,
  paste(labwide$ARM, labwide$Timepoint, sep="."), mean)
```

## Simple t.test

We define logical vectors to help us select the desired data.

```{r,eval=TRUE}
fluad    <- labwide$ARM == "FLUAD"
agrippal <- labwide$ARM == "AGRIPPAL"

d0 <- labwide$Timepoint == "D0"
d1 <- labwide$Timepoint == "D1"
```

## Simple t.test

Paired or not paired?

```{r,eval=TRUE,results="markdown"}
x <- labwide$NEUT
t.test(x[d0 & fluad], x[d1 & fluad])
``` 

## Simple t.test

Paired or not paired?

```{r,eval=TRUE,results="markdown"}
t.test(x[d0 & fluad], x[d1 & fluad], paired=TRUE)
``` 

## Simple t.test

The object returned by `t.test` is a list with the results of the test.

```{r,eval=TRUE,results="markdown"}
res <- t.test(x[d0 & fluad], x[d1 & fluad], paired=TRUE)
res$p.value
```

We can directly insert the p-value into a text using backticks: `` `r
format.pval(res$p.value)` ``.

Result:

We can directly insert the p-value into a text using backticks: `r format.pval(res$p.value)`.


## Broom: cleaning up data

`tidy()` cleans up messy results of various standard statistical tests and
formats them into a nice, clean, TIDY data frame (or, rather, a tibble).

```{r,eval=FALSE,results="markdown"}
library(broom)
tidy(t.test(x[d0 & fluad], x[d1 & fluad], paired=TRUE))
```

## Caveat: paired data

We must make sure that all the data points are paired!

## Exercise 5.2

 * Use the iris dataset
 * Using t-test, check whether the sepal length is different between the species

## PCA

```{r eval=TRUE}
library(tidyverse)
data(iris)
theme_set(theme_bw())
pca <- prcomp(iris[,1:4], scale.=TRUE)
df <- data.frame(pca$x, species=iris$Species)
ggplot(df, aes(x=PC1, y=PC2, color=species)) + geom_point()
```

## PCA (cont.)

```{r,eval=TRUE,echo=FALSE}
data(iris)
theme_set(theme_bw())
pca <- prcomp(iris[,1:4], scale.=TRUE)
df <- data.frame(pca$x, species=iris$Species)
ggplot(df, aes(x=PC1, y=PC2, color=species)) + geom_point()
```


# Summary - Let's bring all concepts together

## General advice

 * Force yourself to use R/Rmarkdown for everything. It will pay off.
 * If it is stupid but it works, it is not stupid.
 * Work with real data rather than examples. If you know what you want to
   do, the learning will be more effective.
 * However, when trying a new package or function, first try it on the
   examples included, then try to use it on your data.
 * Always create a project directory
 * use sessionInfo() to document the environment

## Summary - Let's bring all concepts together

![](images/summaryCourse.png){ width=900px }

## Where to find R packages

* [CRAN repository](https://cran.r-project.org/)
  * packages for statistics, visualization, machine learning, ...
  * see Task Views for a collection of topics 
  * includes also *tidyverse* universe
* [Bioconductor](http://bioconductor.org/)
  * Tools / Packages for different types of bioinformatic analysis, eg
  * *flowStats*, *flowViz*, ...
  * Guides / Books for workflows, eg
  * [Cytof data](http://bioconductor.org/packages/release/workflows/html/cytofWorkflow.html)
* GitHub: many packages are hosted on GitHub

## Where to go from here and where to find help

 * [StackOverflow](https://stackoverflow.com)
 * Walk-in clinic
 * Online R books
   * [Hands-On programming with R](https://rstudio-education.github.io/hopr/)
   * [R for data science (tidyverse)](https://r4ds.had.co.nz/)
   * [R graphics cookbook](https://r-graphics.org/)
   * [R markdown: the definitive guide](https://bookdown.org/yihui/rmarkdown/)
 * Other books:
   * [The R Book](https://www.wiley.com/en-us/The+R+Book%2C+2nd+Edition-p-9780470973929)


